nselib/httpcookies.lua

```diff
@@ -0,0 +1,534 @@
+
+    if ( o:parse(self.cookies) ) then
+      return o
+    end
+
+  end,
+
+
+  --- Parses the cookie and and splits it into its attributes if its a string.
+  -- @param cookies A cookie table 
+  -- @return status
+  -- @return cookies The cookie table in case parsed cookie is a string
+  parse = function(cookies)
+    if cookies == nil then
+      return true
+    end
+    if (type(cookies) == 'table') then
+      --Name and Value must be present in the cookie, rest
+      --attributes are optional.
+      --So, if we have a table of cookies, we take the first table 
+      --and then loop for all the attributes in that table.
+      for _,cookie in ipairs(cookies) do
+        for cookie_attribute, cookie_attribute_value in pairs(cookie) do
+          cookie_attribute = cookie_attribute:lower()
+          if(cookie_attribute == 'name') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie name is not a string")
+              return false --Name has to be of type string
+            end
+          elseif(cookie_attribute == 'value') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie value is not a string")
+              return false
+            end
+          elseif(cookie_attribute == 'path') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie path is not a string")
+              return false
+            end
+          elseif(cookie_attribute == 'expires') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie expires attribute is not a string")
+              return false
+            end
+          elseif(cookie_attribute == 'max-age') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie max-age attribute is not a string")
+              return false
+            end
+          elseif(cookie_attribute == 'domain') then
+            if(type(cookie_attribute_value) ~= 'string') then
+              stdnse.debug1("Cookie domain attribute is not a string")
+              return false
+            end
+          elseif(cookie_attribute == 'secure') then
+            if(type(cookie_attribute_value) ~= 'boolean') then
+              stdnse.debug1("Cookie secure attribute is not a boolean")
+              return false
+            end
+          elseif(cookie_attribute == 'httponly') then
+            if(type(cookie_attribute_value) ~= 'boolean') then
+              stdnse.debug1("Cookie httponly attribute is not a boolean")
+              return false
+            end
+          else
+            stdnse.debug1("Cookie attribute is not recognised")
+            return false
+          end
+        end
+      end
+      return true
+    elseif (type(cookies) == 'string') then
+      --We can parse it using the http parse cookie function
+      cookies = http.parse_set_cookie(cookies)
```

Comment:
Maybe it is the time to refactor all the code related to cookies inside http.lua and move it to httpcookies. Can you look into what scripts are using the cookie functionality and update those scripts to use the new library instead?

Reply:
Well, this function is being used to validate options in every http request. Also, we cant make http to require httpcookies, that would lead to recursive calls. So I think its better like this

Reply:
I think we need to refactor everything related to HTTP cookies and include it in this library. The functions are there because we didn't have the library back then. Now, it doesn't make sense to keep the functions to there. Our next step is to refactor everything that was supporting cookies to use the new library. And this includes old functions living in http.lua or other scripts.

Reply:
So, I checked again. There is some cookie check code in validate_options of http.lua and I think it can stay there since all the options are validated after every http request. I will look into the old scripts(I guess there is http-brute) so that it can use the new library.